src/aggregator/aggregator/handler/writer/sharded.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3/src/aggregator/sharding"
+	"github.com/m3db/m3/src/metrics/metric/aggregated"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/instrument"
+)
+
+var (
+	errShardedWriterNoWriters = errors.New("no backing writers provided")
+	errShardedWriterClosed    = errors.New("sharded writer closed")
+)
+
+type shardedWriter struct {
+	sync.RWMutex
+	closed bool
+
+	numShards int
+	writers   []Writer
+	locks     []sync.Mutex
+	shardFn   sharding.AggregatedShardFn
+}
+
+var _ Writer = &shardedWriter{}
```

Comment:
nit: can we put this with the other `var`s above?

Reply:
So I prefer these interface binding decls to be separate from and right next to the type we're constraining. Makes it a lot more obvious what the intent is. 

Putting them in the same var decl as the other consts feels noisier.